port/osx/omrosbacktrace_impl.c

```diff
@@ -0,0 +1,433 @@
+ *
+ * @return the number of addresses recorded in the array
+ */
+static uintptr_t
+unw_backtrace(void **array, uintptr_t size)
+{
+	unw_cursor_t cursor;
+	unw_cursor_t prev;
+	unw_context_t uc;
+	unw_word_t ip = 0;
+	char buf[64];
+	uintptr_t nAddr = 0;
+
+	unw_getcontext(&uc);
+	unw_init_local(&cursor, &uc);
+	while ((unw_step(&cursor) > 0) && (nAddr < size)) {
+		unw_get_reg(&cursor, UNW_REG_IP, &ip);
+		array[nAddr] = (void *)ip;
+		nAddr += 1;
+
+		/* Save frame for check below. */
+		memcpy(&prev, &cursor, sizeof(unw_cursor_t));
+	}
+
+	/* Darwin's libunwind cannot handle sigtramp in some cases. If the final frame iterated is sigtramp,
+	 * there are more frames but need to manually restore registers.
+	 */
+	if (0 == unw_get_proc_name(&prev, buf, sizeof(buf), &ip)) {
+		if (0 == strcmp(buf, "_sigtramp")) {
+			nAddr = handle_sigtramp(&prev, array, size, nAddr);
+		}
+	}
+
+	return nAddr;
+}
+
+/**
+ * NULL handler. We only care about preventing the signal from propagating up the call stack, no need to do
+ * anything in the handler.
+ *
+ * @param[in] portLibrary a pointer to an initialized port library
+ * @param[in] gpType type of signal
+ * @param[in] gpInfo contains data about signal
+ * @param[in] userData data passed in by user
+ *
+ * @return OMRPORT_SIG_EXCEPTION_RETURN indicating that a signal was received
+ */
+static uintptr_t
+handler(OMRPortLibrary *portLibrary, uint32_t gpType, void *gpInfo, void *userData)
+{
+	return OMRPORT_SIG_EXCEPTION_RETURN;
+}
+
+/**
+ * Wrapped call to libunwind backtrace functions.
+ *
+ * @param[in] port a pointer to an initialized port library
+ * @param[out] arg a pointer to the addresses array and capacity
+ *
+ * @return the number of addresses recorded in the array
+ */
+static uintptr_t
+protectedBacktrace(OMRPortLibrary *port, void *arg)
+{
+	frameData *addresses = (frameData *)arg;
+	return unw_backtrace(addresses->address_array, addresses->capacity);
+}
+
+/**
+ * Provides signal protection for the unw_backtrace call. If the stack walk causes a segfault then we check the output
+ * array to see if we got any frames as we may be able to provide a partial backtrace. We also record that the stack walk
+ * was incomplete so that information is available after the function returns.
+ *
+ * @param[in] portLibrary a pointer to an initialized port library
+ * @param[out] threadInfo the thread structure to hold an error code, if any. Must not be NULL
+ * @param[out] address_array the array to hold the frame addresses
+ * @param[in] capacity capacity of the array
+ *
+ * @return the number of addresses recorded in the array
+ */
+static uintptr_t
+backtrace_sigprotect(OMRPortLibrary *portLibrary, J9PlatformThread *threadInfo, void **address_array, uintptr_t capacity)
+{
+	uintptr_t ret = 0;
+	frameData args;
+	args.address_array = address_array;
+	args.capacity = capacity;
+
+	memset(address_array, 0, sizeof(void *) * capacity);
+
+	if (NULL != omrthread_self()) {
```

Comment:
Suggest inverting this `if` statement, so both branches begin closer to the condition making the code easier to follow.